[BREAKING] Strict conversion (#10223)
ethereum/solidity: docs/080-breaking-changes.rst

  3. Explicit conversions between literals and enums are only allowed if the literal can
     represent a value in the enum.

* There are new restrictions on explicit conversions between two different types. The conversion is

>   only allowed when there is at most one change in sign, width or 'kind' (``int``, ``address``,

Reviewer: I think here you could just say type instead of kind.

  ``bytesNN``, etc.) For example, the conversion ``uint16(int8)`` is disallowed since the conversion
  changes width (8 bits to 16 bits) and sign (signed integer to unsigned integer.) To get the
  previous behaviour, use an intermediate conversion. In the previous example, this would be

>   ``uint16(uint8(int8))``. The following are some examples of conversions that are disallowed by

Reviewer: ```suggestion
  ``uint16(uint8(int8))``, or ``uint16(int16(int8))``. The following are some examples of conversions that are disallowed by
```

  3. Explicit conversions between literals and enums are only allowed if the literal can
     represent a value in the enum.

> * There are new restrictions on explicit conversions between two different types. The conversion is

Reviewer 2: ```suggestion
* There are new restrictions on explicit type conversions. The conversion is
```

  only allowed when there is at most one change in sign, width or type-category (``int``,
  ``address``, ``bytesNN``, etc.) For example, the conversion ``uint16(int8)`` is disallowed since

>   the conversion changes width (8 bits to 16 bits) and sign (signed integer to unsigned integer.) To

Reviewer 2: ```suggestion
  the conversion changes width (8 bits to 16 bits) and sign (signed integer to unsigned integer). To
```

  get the previous behaviour, use an intermediate conversion. In the previous example, this would be

>   ``uint16(uint8(int8))`` or ``uint16(int16(int8))``. The following are some examples of conversions

Reviewer 2: Note that the two ways to convert will produce different results for e.g. `-1`

Author: Added this.

  that are disallowed by this rule. Note that, given types ``T`` and ``S``, the notation ``T(S)``
  refers to the explicit conversion ``T(x)``, where ``x`` is any arbitrary variable of type ``S``.

>   - ``address(uint)`` and ``uint(address)``: converting both 'kind' and width. Replace this by

Reviewer: I still find this `'kind'` weird, especially because it's always single quoted

Author: A problem with using type would be: distinguishing the difference between `uint16` v/s `int8` and 'uint16` v/s `address`. Each pair has different type, but the latter one has different kind. 

But what do you think is a good replacement?

Reviewer 2: type category?

Reviewer: Agree with type category

Author: Replaced 'kind' by type-category.

  3. Explicit conversions between literals and enums are only allowed if the literal can
     represent a value in the enum.

* There are new restrictions on explicit type conversions. The conversion is only allowed when there
  is at most one change in sign, width or type-category (``int``, ``address``, ``bytesNN``, etc.)

>   For example, the conversion ``uint16(int8)`` is disallowed since the conversion changes width (8

Reviewer 2: ```suggestion
  For example, the conversion ``int8`` to ``uint16`` is disallowed since the conversion changes width (8
```

>   bits to 16 bits) and sign (signed integer to unsigned integer). To get the previous behaviour, use

Reviewer 2: ```suggestion
  bits to 16 bits) and sign (signed integer to unsigned integer). In order to do the conversion,
  you have to go through an intermediate type.
```

>   an intermediate conversion. In the previous example, this would be ``uint16(uint8(int8))`` or

Reviewer 2: ```suggestion
  an intermediate conversion. In the previous example, this would be ``int8 -> uint8 -> uint16`` or
```

>   ``uint16(int16(int8))``. Note that the two ways to convert will produce different results e.g.,

Reviewer 2: ```suggestion
  ``int8 -> int16 -> uint16``. Note that the two ways to convert will produce different results e.g.,
```

  for ``-1``. The following are some examples of conversions that are disallowed by this rule. Note

>   that, given types ``T`` and ``S``, the notation ``T(S)`` refers to the explicit conversion

Reviewer 2: We can also use this notation above if you move the explanation up.

Author: Then, I'll just move the explanation up.

  3. Explicit conversions between literals and enums are only allowed if the literal can
     represent a value in the enum.

* There are new restrictions on explicit type conversions. The conversion is only allowed when there
  is at most one change in sign, width or type-category (``int``, ``address``, ``bytesNN``, etc.)

>   Let us use the notation ``T(S)`` to denote the explicit conversion ``T(x)``, where, ``T`` and

Author: I moved the text about documentation and did minor rewordings.

I'll update the conversions on fixed point types.

  ``S`` are types, and ``x`` is any arbitrary variable of type ``S``. The conversion

>   ``uint16(int8)`` is now disallowed since the conversion changes width (8 bits to 16 bits) and sign

Reviewer 2: Can you make it clear that this is just an example? It could be read as the main change.

Author: Updated.
